src/cargo/core/compiler/compilation.rs

```diff
@@ -116,11 +119,13 @@ impl<'cfg> Compilation<'cfg> {
                 .build_config
                 .requested_kinds
                 .iter()
-                .chain(Some(&CompileKind::Host))
+                .copied()
+                .chain(Some(CompileKind::Host))
+                .chain(ws.members().flat_map(|p| p.manifest().kind().into_iter()))
```

Comment:
Similar to above, I think this needs to iterate over the entire crate graph?

(this may indicate that this should be refactored to a separate step where only one location determines all the targets in play and then that's passed as an argument here or stored in `bcx`)